src/cpp/src/multinomial_decoding.cpp

```diff
@@ -0,0 +1,271 @@
+    float top_p;
+    float temperature;
+    float repetition_penalty;
+
+    SamplingParameters(ov::GenerationConfig generation_config) {
+        // parameters validation
+        OPENVINO_ASSERT(generation_config.top_k > 0,
+                        "top_k must be a strictly positive float, but got ",
+                        generation_config.top_p);
+        OPENVINO_ASSERT(generation_config.top_p > 0 || generation_config.top_p < 1.0f,
+                        "top_p must be a positive float > 0 and < 1, but got ",
+                        generation_config.top_p);
+        OPENVINO_ASSERT(generation_config.temperature > 0,
+                        "Temperature must be a strictly positive float, but got ",
+                        generation_config.temperature);
+        OPENVINO_ASSERT(generation_config.repetition_penalty > 0,
+                        "Repetition penalty must be a strictly positive float, but got ",
+                        generation_config.repetition_penalty);
+
+        top_k = generation_config.top_k;
+        top_p = generation_config.top_p;
+        temperature = generation_config.temperature;
+        repetition_penalty = generation_config.repetition_penalty;
+    }
+};
+
+struct RandomSampling {
+    SamplingParameters parameters;
+    RandomSampling(SamplingParameters parameters) : parameters{std::move(parameters)} {}
+
+    TokenIdScore get_out_token(float* logits, size_t vocab_size, std::vector<int64_t> tokens) {
+        // logits pre-process
+        if (parameters.repetition_penalty != 1.0f) {
+            sampling_repetition_penalty(logits, logits + vocab_size, tokens, parameters.repetition_penalty);
+        }
+
+        if (parameters.temperature > 0) {
+            sampling_temperature(logits, logits + vocab_size, parameters.temperature);
+        }
+
+        std::vector<TokenIdScore> token_scores(vocab_size);
+        for (int i = 0; i < vocab_size; i++) {
+            token_scores[i] = TokenIdScore(i, logits[i]);
+        }
+
+        // top_k sampling
+        if (0 < parameters.top_k && parameters.top_k < (int)token_scores.size()) {
+            sampling_top_k(token_scores.data(),
+                           token_scores.data() + parameters.top_k,
+                           token_scores.data() + token_scores.size());
+            token_scores.resize(parameters.top_k);
+        }
+
+        // top_p sampling
+        if (0.f < parameters.top_p && parameters.top_p < 1.0f) {
+            auto pos = sampling_top_p(token_scores.data(), token_scores.data() + token_scores.size(), parameters.top_p);
+            token_scores.resize(pos - token_scores.data());
+        }
+
+        // sample next token
+        sampling_softmax_inplace(token_scores.data(), token_scores.data() + token_scores.size());
+        for (size_t i = 0; i < token_scores.size(); i++) {
+            logits[i] = token_scores[i].score;
+        }
+
+        thread_local std::random_device rd;
+        thread_local std::mt19937 gen(rd());

```

Comment:
I don't have permissions to resolve conversations in this fork. Consider the conversations I didn't reply in as resolved.

`thread_local` must be global to take an effect:
> a term used to refer to data that is seemingly global or static storage duration (from the viewpoint of the functions using it)

You can try running the example from the second best answer there. Declaring `i` as `thread_local` prints a deterministic set of numbers, while the order is random. Removing `thread_local` results in possible print of the same digit twice, for example `3244`.

I came up with an example similar to your usecase:
```C++
#include <iostream>
#include <thread>

void thread_func(int id)
{
    thread_local int i = 0;
    i = id;
    ++i;
    std::cout << i << '\n';
}

int main()
{
    std::thread t1(thread_func, 1);
    std::thread t2(thread_func, 2);
    std::thread t3(thread_func, 3);

    t1.join();
    t2.join();
    t3.join();

}
```
Keeping or removing `thread_local` doesn't affect the behavior.

It may be a good idea to submit a ticket to the original repo why they use `thread_local`.

Anyway, we are going to allow seed configuration sooner or later: https://github.com/openvinotoolkit/openvino.genai/pull/334#discussion_r1605964450, so the generator will end up being a struct member anyway. Please, move it to class member.

Reply:
Moved to class member